zebra-chain/src/sapling/shielded_data.rs

```diff
@@ -216,40 +257,51 @@ where
     }
 }
 
-// Technically, it's possible to construct two equivalent representations
-// of a ShieldedData with at least one spend and at least one output, depending
-// on which goes in the `first` slot.  This is annoying but a smallish price to
-// pay for structural validity.
-//
-// A `ShieldedData<PerSpendAnchor>` can never be equal to a
-// `ShieldedData<SharedAnchor>`, even if they have the same effects.
-
-impl<AnchorV> std::cmp::PartialEq for ShieldedData<AnchorV>
+impl<AnchorV> TransferData<AnchorV>
 where
-    AnchorV: AnchorVariant + Clone + PartialEq,
+    AnchorV: AnchorVariant + Clone,
 {
-    fn eq(&self, other: &Self) -> bool {
-        // First check that the lengths match, so we know it is safe to use zip,
-        // which truncates to the shorter of the two iterators.
-        if self.spends().count() != other.spends().count() {
-            return false;
-        }
-        if self.outputs().count() != other.outputs().count() {
-            return false;
+    /// Iterate over the [`Spend`]s for this transaction, returning them as
+    /// their generic type.
+    pub fn spends(&self) -> impl Iterator<Item = &Spend<AnchorV>> {
+        use TransferData::*;
+
+        let spends = match self {
+            SpendsAndMaybeOutputs { spends, .. } => Some(spends.iter()),
+            JustOutputs { .. } => None,
+        };
+
+        // this awkward construction avoids returning a newtype struct or
+        // type-erased boxed iterator
+        spends.into_iter().flatten()
+    }
+
+    /// Iterate over the [`Output`]s for this transaction.
+    pub fn outputs(&self) -> impl Iterator<Item = &Output> {
+        use TransferData::*;
```

Comment:
If we're importing this inside several methods, should we just import it at the top of the module?

Reply:
So this is a bit of a weird one, because `TransferData` is an enum.

Sometimes you want the fully qualified name, if it's not obvious from the context.